extensions/amp-ad-network-fake-impl/0.1/amp-ad-metadata-transformer.js

```diff
@@ -0,0 +1,300 @@
+    }
+    for (let i = 0; i < head.children.length; i++) {
+      const child = head.children.item(i);
+      if (child.tagName == 'SCRIPT' && child.hasAttribute('src')) {
+        if (this.firstRuntimeElement_ == null) {
+          this.firstRuntimeElement_ = child;
+        }
+        this.lastRuntimeElement_ = child;
+        if (child.hasAttribute('custom-element')) {
+          this.extensions_.push({
+            'custom-element': child.getAttribute('custom-element'),
+            'src': child.getAttribute('src'),
+          });
+        }
+        if (child.hasAttribute('custom-template')) {
+          this.extensions_.push({
+            'custom-template': child.getAttribute('custom-template'),
+            'src': child.getAttribute('src'),
+          });
+        }
+      }
+      if (
+        child.tagName == 'LINK' &&
+        child.getAttribute('rel') == 'stylesheet' &&
+        child.hasAttribute('href')
+      ) {
+        this.generateStyleMetadata_(child);
+      }
+      if (
+        child.tagName == 'META' &&
+        child.hasAttribute('name') &&
+        child.hasAttribute('content')
+      ) {
+        this.generateCtaMetadata_(child);
+      }
+    }
+  }
+
+  /**
+   * Generates custom style metadata
+   * @param {Element} element
+   */
+  generateStyleMetadata_(element) {
+    const styleJson = {href: element.getAttribute('href')};
+    if (element.hasAttribute('media')) {
+      styleJson.media = element.getAttribute('media');
+    }
+    this.styles_.push(styleJson);
+  }
+
+  /**
+   * Generates CTA metadata for story ads
+   * @param {Element} element
+   */
+  generateCtaMetadata_(element) {
+    if (element.getAttribute('name') == 'amp-cta-type') {
+      this.ctaType_ = element.getAttribute('content');
+    }
+    if (element.getAttribute('name') == 'amp-cta-url') {
+      this.ctaUrl_ = element.getAttribute('content');
+    }
+  }
+
+  /**
+   * Generates image metadata
+   * @param {Document} doc
+   */
+  generateImageMetadata_(doc) {
+    const imgs = doc.querySelectorAll('amp-img[src]');
+    for (let i = 0; i < imgs.length; i++) {
+      const img = imgs[i];
+      let width,
+        height,
+        area = -1;
+      if (img.hasAttribute('width')) {
+        width = img.getAttribute('width');
+      }
+      if (img.hasAttribute('height')) {
+        height = img.getAttribute('height');
+      }
+      if (height && width) {
+        area = height * width;
+      }
+      this.imgMetadata_.push({
+        src: img.getAttribute('src'),
+        area,
+      });
+    }
+  }
+
+  /**
+   * Finds all <script type=application/json> tags and parses
+   * existing <amp-ad-metadata> script, if it exists
+   * @param {Document} doc
+   */
+  generateJsonMetadata_(doc) {
+    const json = doc.querySelectorAll('script[type]');
+    for (let i = 0; i < json.length; i++) {
+      const script = json[i];
+      const type = script.getAttribute('type');
+      if (type != 'application/json') {
+        return;
```

Comment:
Talked offline, please add warning that if we get an invalid amp doc here it may have weird results. Also please add test that checks that warning happens.

Reply:
Done